[cmake] prevent "@ADDON_DEPENDS@" with same several times

AlwinEsch opened the pull request:
## Description

Before comes e.g.:
```
-- Added usage definition: ADDON_GLOBAL_VERSION_MAIN_USED
-- Added usage definition: ADDON_GLOBAL_VERSION_GUI_USED
-- Added usage definition: ADDON_GLOBAL_VERSION_GUI_USED <<<<<<<<<<<<<<<<<<<<<<< THIS HERE
-- Added usage definition: ADDON_INSTANCE_VERSION_SCREENSAVER_USED
```

or on addon.xml:

```xml
  <requires>
<import addon="kodi.binary.global.main" version="1.0.15"/>
<import addon="kodi.binary.global.gui" version="5.13.0"/>
<import addon="kodi.binary.global.gui" version="5.13.0"/> <<<<<<<<<<<<<<<<<<<<<<< THIS HERE
<import addon="kodi.binary.instance.screensaver" version="2.0.0"/></requires>
```

This adds a test to whether this definition has been previously set.

<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->

## How Has This Been Tested?
On Binary addon, after them is only:
```xml
  <requires>
<import addon="kodi.binary.global.main" version="1.0.15"/>
<import addon="kodi.binary.global.gui" version="5.13.0"/>
<import addon="kodi.binary.instance.screensaver" version="2.0.0"/></requires>
```

This does not create a version problem on addons, only with newly created this double entry is avoided.

<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

AlwinEsch commented:
Have added a second commit to fix a include with spaces on begin

Rechi commented on cmake/scripts/common/AddonHelpers.cmake:
```diff
@@ -120,16 +120,20 @@ macro (build_addon target prefix libs)
+                    # Inform with them the addon header about used type, if not present before
+                    get_directory_property(CURRENT_DEFS COMPILE_DEFINITIONS)
+                    if(NOT CURRENT_DEFS MATCHES "${used_type_name}")
+                      add_definitions(-D${used_type_name})
```
What is that needed for?

AlwinEsch replied:
The `add_definitions` is needed for `#if` here: https://github.com/xbmc/xbmc/blob/master/xbmc/addons/kodi-addon-dev-kit/include/kodi/versions.h#L178-L256.

Where addon returns `"0.0.0"` if not needed and becomes ignored from Kodi.

Rechi replied:
Only greped for `ADDON_GLOBAL_VERSION_MAIN_USED` and that didn't show anything.

Rechi reviewed (dismissed):
This doesn't look like it fixes the multiple `import` tags with the same `addon` attribute.

AlwinEsch commented:
> This doesn't look like it fixes the multiple `import` tags with the same `addon` attribute.

Have update the `if`, was a line to low and addon.xml has still included it,

Rechi commented:
Not using the preprocessor, before 'parsing' the source files is error prone.
E.g. includes which are commented or ifdefed are considered as uses too.

Rechi commented on cmake/scripts/common/AddonHelpers.cmake:
```diff
@@ -120,16 +120,20 @@ macro (build_addon target prefix libs)
-                    add_definitions(-D${used_type_name})
-                    message(STATUS "Added usage definition: ${used_type_name}")
+                    get_directory_property(CURRENT_DEFS COMPILE_DEFINITIONS)
+                    if(NOT CURRENT_DEFS MATCHES "${used_type_name}")
```
`MATCHES` is wrong as it would also match definitions which contain `${used_type_name}`, but have additional characters at the beginning or end.

AlwinEsch replied:
Not know direct a good alternative.
e.g. on Linux as definition it have `BUILD_KODI_ADDON;HAS_GL=1;ADDON_GLOBAL_VERSION_MAIN_USED;ADDON_GLOBAL_VERSION_GUI_USED;ADDON_INSTANCE_VERSION_SCREENSAVER_USED`, but not sure if this is on all OS like them.

But do not look directly dangerous as it always has this at the beginning and end:
- `ADDON_*`
- `*_USED`

Rechi replied:
```suggestion
                    if(NOT used_type_name IN_LIST CURRENT_DEFS)
```

AlwinEsch replied:
Thanks a lot! This `IN_LIST` is new for me, but it works good here.

With the "not sure" I meant before that mostly in `add_definitions()` a `-D ...` is to begin, and thought that this may not always be a list, but actually only makes sense as a list.

AlwinEsch committed:
[cmake] prevent "@ADDON_DEPENDS@" with same several times

Before comes e.g.:
```
-- Added usage definition: ADDON_GLOBAL_VERSION_MAIN_USED
-- Added usage definition: ADDON_GLOBAL_VERSION_GUI_USED
-- Added usage definition: ADDON_GLOBAL_VERSION_GUI_USED <<<<<<<<<<<<<<<<<<<<<<< THIS HERE
-- Added usage definition: ADDON_INSTANCE_VERSION_SCREENSAVER_USED
```

or on addon.xml:

```xml
  <requires>
<import addon="kodi.binary.global.main" version="1.0.15"/>
<import addon="kodi.binary.global.gui" version="5.13.0"/>
<import addon="kodi.binary.global.gui" version="5.13.0"/> <<<<<<<<<<<<<<<<<<<<<<< THIS HERE
<import addon="kodi.binary.instance.screensaver" version="2.0.0"/></requires>
```

This adds a test to whether this definition has been previously set.

AlwinEsch commented:
Does it also make sense to make a request to Leia? This concerns it too.

Rechi commented on cmake/scripts/common/AddonHelpers.cmake:
```diff
@@ -120,16 +120,20 @@ macro (build_addon target prefix libs)
             foreach(src_file ${USED_SOURCES})
               file(STRINGS ${src_file} BIN_ADDON_SRC_PARTS)
               foreach(loop_var ${BIN_ADDON_SRC_PARTS})
+                string(REGEX REPLACE ".*#include" "#include" loop_var "${loop_var}")
```
This `REGEX REPLACE` is error prone, as it would match `#includes` which have other characters the spaces at the begin. Also it doesn't take care about spaces after the `#` or using `#import` (Objective-C++).

AlwinEsch replied:
Have added a space handling after `#` and also tanslate of `#import`

Rechi commented on cmake/scripts/common/AddonHelpers.cmake:
```diff
@@ -120,16 +120,20 @@ macro (build_addon target prefix libs)
                 string(FIND "${loop_var}" "#include" matchres)
                 if("${matchres}" EQUAL 0)
                   string(REPLACE " " ";" loop_var "${loop_var}")
                   list(GET loop_var 1 include_name)
```
This is wrong if `#include` is followed by more then once space.

AlwinEsch replied:
No, is correct for them is the `*` after `.`

Rechi replied:
Tested with `#include  <kodi/addon-instance/Screensaver.h>` and it's not detected.

AlwinEsch replied:
Have it checked in cmake regex, unfortunately, I did not find a usable way directly.
It is possible to use regex groups in output (`\\1`), but unfortunately if the group is missing, it's a bit crazy,

Found some options with https://regex101.com unfortunately, but cmake does not support this.

With this request it is still not perfect, but at least now more is found.

Rechi replied:
You implement this in a very complicated way, regex replace, find, replace, select n-th element and regex replace instead of a single regex match.

AlwinEsch replied:
Found a bit better way (last commit).

Old:
```cmake
                  string(REPLACE " " ";" loop_var "${loop_var}")
                  list(GET loop_var 1 include_name)
                  string(REGEX REPLACE "[<>\"]|kodi/" "" include_name "${include_name}")
```

New:
```cmake
                  string(REGEX REPLACE "#include|[ <>\"]|kodi/" "" include_name "${loop_var}")
```

Also handle this all the spaces between `#include` and `<` and only one line.

With the space there, unfortunately, one was removed within <>, but since it refers to Kodi header check, which never have spaces in it, I do not see it bad.

Rechi reviewed (commented):
Keep in mind that `if(<variable|string> IN_LIST <variable>)` requires every add-on to have at least `cmake_minimum_required(VERSION 3.3)` set in it's main `CMakeLists.txt`.

AlwinEsch commented:
Acceptable to bring in?

Rechi commented on cmake/scripts/common/AddonHelpers.cmake:
```diff
@@ -120,16 +120,18 @@ macro (build_addon target prefix libs)
             foreach(src_file ${USED_SOURCES})
               file(STRINGS ${src_file} BIN_ADDON_SRC_PARTS)
               foreach(loop_var ${BIN_ADDON_SRC_PARTS})
+                string(REGEX REPLACE ".*#*.*include|.*#*.*import" "#include" loop_var "${loop_var}")
```
This is wrong as it removes all characters before `#` and between `#` and `include` or `import`, instead of removing only white space characters.

Rechi commented on cmake/scripts/common/AddonHelpers.cmake:
```diff
@@ -120,16 +120,18 @@ macro (build_addon target prefix libs)
-                  string(REPLACE " " ";" loop_var "${loop_var}")
-                  list(GET loop_var 1 include_name)
-                  string(REGEX REPLACE "[<>\"]|kodi/" "" include_name "${include_name}")
+                  string(REGEX REPLACE "#include|[ <>\"]|kodi/" "" include_name "${loop_var}")
```
This removes spaces also in the filename and doesn't care about tabs.

Rechi commented on cmake/scripts/common/AddonHelpers.cmake:
```diff
@@ -120,16 +120,18 @@ macro (build_addon target prefix libs)
-                  list(GET loop_var 1 include_name)
-                  string(REGEX REPLACE "[<>\"]|kodi/" "" include_name "${include_name}")
+                  string(REGEX REPLACE "#include|[ <>\"]|kodi/" "" include_name "${loop_var}")
                   if(include_name MATCHES ${depend_header})
```
This actually needs to be a starts with and not a contains anywhere checks.

Rechi reviewed (commented):
This is still too much string fiddling which isn't easily understandable.
The whole parsing can be done with a single `string(REGEX MATCH ...)` in CMake 3.5. My previous edit about not being able to use that was wrong, it has the `CMAKE_MATCH_(0..9)` variables too. It's just a bit hidden in the documentation of older versions.

AlwinEsch commented:
Is updated and now only with one big regex line.
Thought also to use:
```
^[ \t]*#*[ \t](?:include|import)[ \t]*[\"<].*[\">]
```
where more clear, but cmake not work with "(?:include|import)" check inside.

AlwinEsch commented:
I can also use a:
```
string(REGEX MATCH "^[ \t]*#*[ \t]*import*[ \t]*[\"<]kodi/${depend_header}.*[\">]|^[ \t]*#*[ \t]*include*[ \t]*[\"<]kodi/${depend_header}.*[\">]" loop_var "${loop_var}")
if(loop_var)
  ...
```

without the `if(loop_var AND loop_var MATCHES "kodi/${depend_header}")`, but I'm not sure if it would be good because you can see what it is all about.

Do you know another way to shorten this long regex thing? Like a on cmake working `(?:include|import)`?

Rechi commented on cmake/scripts/common/AddonHelpers.cmake:
```diff
@@ -120,18 +120,16 @@ macro (build_addon target prefix libs)
-                  list(GET loop_var 1 include_name)
-                  string(REGEX REPLACE "[<>\"]|kodi/" "" include_name "${include_name}")
-                  if(include_name MATCHES ${depend_header})
+                string(REGEX MATCH "^[ \t]*#*[ \t]*import*[ \t]*[\"<].*[\">]|^[ \t]*#*[ \t]*include*[ \t]*[\"<].*[\">]" loop_var "${loop_var}")
```
`#`, `t` and `e` are allowed zero to unlimited times

Rechi commented:
`(?:include|import)` -> `(include|import)`

AlwinEsch commented:
> `(?:include|import)` -> `(include|import)`

Ah, Thanks! Seen this before as group separation, but it works.

Is updated again and also now with one `#` only to if.

AlwinEsch commented:
@Rechi is it OK now?

jenkins build this now

AlwinEsch commented:
jenkins build this please

AlwinEsch commented:
@Rechi "regex" and changes acceptable?

AlwinEsch commented:
jenkins build this please

AlwinEsch commented:
> note: this breaks currently the check by old addon type (PVR & Game) where no "<kodi/..." is used on start.

Rechi commented:
Then adjust the regex, so that `kodi/` is optional again.

AlwinEsch commented:
Possible yes, but there are files like "General.h" or "Network.h" where makes it hard, this name can also be easy on addon itself. A `kodi/` can not be optional by this name of Header files.

From my side we must wait until https://github.com/xbmc/xbmc/pull/16485 and https://github.com/xbmc/xbmc/pull/16453 are in

AlwinEsch commented:
Moment, without this request is it also not completely correct.

I remove it temporary but after the other requests are in, comes the `kodi/` back.

Rechi commented:
Use https://github.com/xbmc/xbmc/compare/master...Rechi:cmake/fixAddonsUsedComponentsParsing and remove the optional `kodi/` once it is no longer optional.

Rechi commented on cmake/scripts/common/AddonHelpers.cmake:
```diff
@@ -120,18 +120,17 @@ macro (build_addon target prefix libs)
-                  string(REGEX REPLACE "[<>\"]|kodi/" "" include_name "${include_name}")
-                  if(include_name MATCHES ${depend_header})
+                string(REGEX MATCH "^[ \t]*#[ \t]*(include|import)[ \t]*[\"<].*[\">]" loop_var "${loop_var}")
+                if(loop_var AND loop_var MATCHES "${depend_header}") # WARNING: Temporary until PVR and Game addons are changed to new C++ style
```
This is wrong, because it doesn't check for an exact match inside `"..."` or `<...>`.

AlwinEsch replied:
I change but there comes also another problem with `if(NOT used_type_name IN_LIST CURRENT_DEFS)`

e.g. pvr.demo still declare: `cmake_minimum_required(VERSION 2.6)`
where generates the `Policy CMP0057 is not set:` where it need `3.3`.

Before must all the PVR CMakeLists.txt updated
Like you already sayed :smile::
> Keep in mind that if(<variable|string> IN_LIST <variable>) requires every add-on to have at least cmake_minimum_required(VERSION 3.3) set in it's main CMakeLists.txt.

I think then to continue my huge addon update chain, now also to PVR. Before I continue here.

AlwinEsch commented:
Updated and all addons should be converted to cmake 3.5.

AlwinEsch commented:
@Rechi acceptable to bring in?

All addons are changed on cmake to 3.5 to support this.

AlwinEsch commented:
jenkins build this with addons please

AlwinEsch commented:
jenkins build this with addons please

Rechi commented on cmake/scripts/common/AddonHelpers.cmake:
```diff
@@ -121,18 +121,16 @@ macro (build_addon target prefix libs)
-                  string(REGEX REPLACE "[<>\"]|kodi/" "" include_name "${include_name}")
-                  if(include_name MATCHES ${depend_header})
+                string(REGEX MATCH "^[ \t]*#[ \t]*(include|import)[ \t]*[\"<].*[\">]" loop_var "${loop_var}")
+                if(loop_var AND loop_var MATCHES "[\"<]kodi/${depend_header}[\">]")
```
```suggestion
                string(REGEX MATCH "^[ \t]*#[ \t]*(include|import)[ \t]*[<\"](kodi/)?(.+)[\">]" include_name ${loop_var})
                if(include_name AND CMAKE_MATCH_3 MATCHES ^${depend_header})
```

AlwinEsch replied:
With this change is in conflict with our Jenkins, no idea why but, it seems to give several times not the right "MATCH" and some depends are not added and mainly `ADDON_GLOBAL_VERSION_MAIN_USED` is missing which is mandatory.

Rechi commented on cmake/scripts/common/AddonHelpers.cmake:
```diff
@@ -121,18 +121,16 @@ macro (build_addon target prefix libs)
                     message(STATUS "Added usage definition: ${used_type_name}")
-                    set(FOUND_HEADER_USAGE 1)
                   endif()
+                  set(FOUND_HEADER_USAGE 1)
```
Why move setting `FOUND_HEADER_USAGE`?

AlwinEsch replied:
Is not moved, only a bit bad presented from git, look the two patches alone and you see. But I move it up to mesage, let see what come.

AlwinEsch commented:
jenkins build this with addons please

AlwinEsch commented:
jenkins build this with addons please

Crrrh, with unknown reason is by Jenkins mostly the `Added usage definition: ADDON_GLOBAL_VERSION_MAIN_USED` missing :unamused: at home it is OK.
I searching now why, sorry if the next time come some of this jenkins comments.

AlwinEsch commented:
Found maybe why, it gives here e.g. https://github.com/xbmc/xbmc/blob/master/xbmc/addons/kodi-addon-dev-kit/include/kodi/versions.h#L40 not the full path.

Possible that a cmake can on a some Version not interpret the `"addon-instance/"` right and it normally for his compare a `\/` for his slash expect?

AlwinEsch commented:
What is the status of this request for you @Rechi , with the change you specified cannot be used. In the current still it works.

AlwinEsch commented:
jenkins build this with addons please

Rechi commented:
What does not work with my suggestion?

AlwinEsch commented:
The Regex there does not find the `addon-instance/`(ADDON_GLOBAL_VERSION_MAIN_DEPENDS)  and `gui/` (ADDON_GLOBAL_VERSION_GUI_DEPENDS) where only a begin of path and not complete.

The addon creation works nice, but during try on Kodi comes a error with a Instance Version mismatch, the required `<import addon="kodi.binary.global.main" minversion="1.0.12" version="1.1.0"/>` then missing.

AlwinEsch commented:
Need to look further also in them now are some addons not correctly set with `ADDON_GLOBAL_VERSION_GENERAL_USED`

AlwinEsch commented:
jenkins build this with addons please

AlwinEsch committed:
[cmake] fix check of " #include <...>" with spaces on begin

There are addons where added the include by `#ifdef` and spaces before
`#include`.

e.g. screensaver.asteroids with:
```cpp
\#ifndef WIN32
  #include <kodi/gui/gl/GL.h>
  #include <kodi/gui/gl/Shader.h>
  #include <glm/glm.hpp>
  #include <glm/gtc/type_ptr.hpp>
\#else
  #include <d3d11.h>
\#endif
```

AlwinEsch commented:
jenkins build this with addons please

AlwinEsch committed:
[cmake] add message to find easier the checks on log

AlwinEsch commented:
jenkins build this with addons please

@Rechi is updated with your suggestion was only a add of `\` and on `${loop_var}` a `"..."` needed to become our Jenkins happy.

With last commit is a message added, this makes it easier to fly over the Jenkins log from build with all addons.

AlwinEsch commented:
Have Jenkins inspected, all addons (for Kodi Matrix) have the correct usage definition with this request and seems OK to merge :smile: 

By a log search with "Addon dependency check" is everything seen.

AlwinEsch committed:
[cmake] show API version by addon dependency check

AlwinEsch commented:
Added also a commit to see used version:
```
-- Addon dependency check ...
--  - Added API usage definition: ADDON_GLOBAL_VERSION_MAIN_USED (Version: "1.1.0", Min. Version: "1.0.12")
--  - Added API usage definition: ADDON_INSTANCE_VERSION_INPUTSTREAM_USED (Version: "2.0.11", Min. Version: "2.0.7")
```
seems better to see direct on Jenkins build which API version is set to the addon

AlwinEsch commented:
jenkins build this with addons please

AlwinEsch commented:
jenkins build this with addons please

AlwinEsch commented:
Confirmed on all 78 by Jenkins created addons the right use and correct set of mandatory "ADDON_GLOBAL_VERSION_MAIN_USED"
